src/server/services/procedures/iotscape/iotscape-services.js

```diff
@@ -0,0 +1,294 @@
+const logger = require('../utils/logger')('iotscape-services');
+
+/**
+ * Stores information about registered services, with a list of IDs and their respective hosts
+ */
+const IoTScapeServices = {};
+
+IoTScapeServices._services = {};
+IoTScapeServices._serviceDefinitions = {};
+
+/**
+ * Creates or updates the connection information for a remote service
+ * @param {String} name 
+ * @param {String} id 
+ * @param {RemoteInfo} rinfo 
+ */
+IoTScapeServices.updateOrCreateServiceInfo = function (name, definition, id, rinfo) {
+    let service = IoTScapeServices._services[name];
+    IoTScapeServices._serviceDefinitions[name] = definition;
+    
+    logger.log('Discovering ' + name + ':' + id + ' at ' + rinfo.address + ':' + rinfo.port);
+    
+    if (!service) {
+        // Service not added yet
+        service = IoTScapeServices._services[name] = {};
+    }
+
+    service[id] = rinfo;
+};
+
+/**
+ * Remove a device from a service
+ * @param {String} service Name of service
+ * @param {String} id ID of device to remove
+ */
+IoTScapeServices.removeDevice = function(service, id) {
+    if(!IoTScapeServices.deviceExists(service, id)){
+        return;
+    }
+
+    delete IoTScapeServices._services[service][id];
+
+    if(IoTScapeServices._listeningClients[service] !== undefined && IoTScapeServices._listeningClients[service][id] !== undefined){
+        delete IoTScapeServices._listeningClients[service][id];
+    }
+};
+
+/**
+ * List IDs of devices associated for a service
+ * @param {String} service Name of service to get device IDs for
+ */
+IoTScapeServices.getDevices = function (service) {
+    if(Object.keys(IoTScapeServices._services).includes(service)){
+        return Object.keys(IoTScapeServices._services[service]);
+    }
+
+    return false;
```

Comment:
I ended up going for the `Error`, I think it makes sense to prevent actions if the service doesn't exist at all (since there's not much the student can do with a non-existent service).